### ISSUE TITLE
Support for AWS Secrets Manager in schema.prisma

### ISSUE THREAD

**Issue by jquirossoto** [prisma/prisma#7534]
## Problem

I'm using Prisma only to manage my database schema, not for the CRUD operations in code. The problem is that the schema.prisma file only supports defining the datasource URL using values defined in a file (dotenv). Since I don't want to store the database credentials in my project I'd like to obtain them from AWS Secrets Manager service.

## Suggested solution

Provide a mechanism like a callback function to defined the datasource URL.

**Comment by janpio** [comment]
That is very reasonable. Do you have an implementation of this in mind or examples how other tools solve this?

**Comment by jquirossoto** [comment]
> That is very reasonable. Do you have an implementation of this in mind or examples how other tools solve this?

I'd imagine something like this:

`datasource db {
  provider = "postgresql"
  url      = await getDatabaseUrl();
}

function getDatabaseUrl() {
  // obtain credentials from cloud
  // create database url based on obtained credentials
  // return databaseUrl;
}`

**Comment by janpio** [comment]
Ok, so being able to call out to a `getDatabaseUrl.js` or `.ts` that executes the logic to return the credential would work for you?

**Comment by jquirossoto** [comment]
Yes that works.

**Comment by janpio** [comment]
Can you share the script and how you called it as a current workaround? That could help @jquirossoto a bit - as we will not implement this tomorrow obviously but more long term.

**Comment by Sytten** [comment]
I dont have access to it, but here is an exemple using GCP SM that is similar:
```javascript
const fs = require('fs');
const path = require('path');
const util = require('util');
const { GoogleAuth } = require('google-auth-library');
const { SecretManagerServiceClient } = require('@google-cloud/secret-manager');

const writeFile = util.promisify(fs.writeFile);
const mkdir = util.promisify(fs.mkdir);
const auth = new GoogleAuth();
const client = new SecretManagerServiceClient();

const SECRETS = {
  DATABASE_URL: (env) => `myproject-${env}-database-url`,
};

const ENVIRONMENTS = {
  production: 'prod',
  staging: 'staging',
};

const getSecretValue = async (name) => {
  const projectId = await auth.getProjectId();
  const secretVersionName = `projects/${projectId}/secrets/${name}/versions/latest`;

  const [secret] = await client.accessSecretVersion({
    name: secretVersionName,
  });

  const payload = secret.payload.data.toString('utf8');
  return payload;
};

async function main() {
  const env = ENVIRONMENTS[process.env.NODE_ENV];

  if (!env) {
    throw new Error('Unsupported environment');
  }

  // Fetch secrets
  const lines = [];
  for (const [localName, gcpName] of Object.entries(SECRETS)) {
    const secretName = gcpName(env);
    const secretValue = await getSecretValue(secretName);
    lines.push(`${localName}=${secretValue}`);
  }

  // Write file
  const content = lines.join('\n');
  const envPath = path.join(__dirname, '..', 'config', 'production.env');

  await mkdir(path.dirname(envPath), { recursive: true });
  await writeFile(envPath, content);
}

main().catch((err) => {
  console.log(err);
  process.exit(1);
});
```

This works well for containers were the `production.env` is discarded when it dies.

For your need I would do something like:
```
DATABASE_URL=$(gcloud secrets versions access latest --secret some-secret) prisma migrate
DATABASE_URL=$(aws secretsmanager get-secret-value --secret-id some-secret --query SecretString --output text) prisma migrate
```
You can even wrap that in a `package.json` script.

**Comment by JameelKhan9** [comment]
@janpio was there any progress or further insight on this issue? We are looking at using the secrets manager similarly and just wanted to check that @Sytten's approach is still the workaround one would be using.

**Comment by janpio** [comment]
No update or it would be mentioned in this issue (or the issue would optimally be closed).

**Comment by aaronstarkey** [comment]
+1. blocked on the same issue. need to be able to use this with docker as well as keeping secrets in the secrets manager. pretty frustrated.

**Comment by shadoath** [comment]
Also looking for this kind of an update to Prisma

**Comment by yogarasu** [comment]
I am also working on this issue,
Is there a solution to this problem?

**Comment by shadoath** [comment]
We ended up with a hacky solution for docker and Prisma.  Our Docker file copies over a shell script: `db_wrapper.sh` that we then execute before running our nextjs app: `CMD ["/app/db_wrapper.sh", "npm", "start"]`

```
#!/usr/bin/env sh

export DATABASE_URL="postgresql://$PGUSER:$PGPASSWORD@$PGHOST:$PGPORT/$PGDATABASE"
exec "$@"
```

We are using the aws cdk to pass in the secrets:
```
...
new PEcsService(appStack, "PEcsService", {
  serviceName: "PEcsService",
  healthCheckPath: "/",
  replicaCount: 1,
  subdomain: "enrollments",
  container: {
    containerName: "EcsContainer",
    image: ecs.ContainerImage.fromAsset(path.join(__dirname, "..", "..")),
    portMappings: [{ containerPort: 3000 }],
    environmentFiles: [],
    environment: {
      PGDATABASE: "ecs",
    },
    secrets: {
      PGHOST: ecs.Secret.fromSecretsManager(db.secret, "host"),
      PGPORT: ecs.Secret.fromSecretsManager(db.secret, "port"),
      PGUSER: ecs.Secret.fromSecretsManager(db.secret, "username"),
      PGPASSWORD: ecs.Secret.fromSecretsManager(db.secret, "password"),
    },
  },
})
...

**Comment by yogarasu** [comment]
Thank you
I was just reaching the same solution
Migration is fine, but I'm looking at the cause because an error occurs in prisma studio.

**Comment by rennehir** [comment]
This feature would be awesome and help us with setting up local environments as well as CI/CD. Also this would help us creating the Prisma client in a serverless GraphQL environment as well.

Meanwhile, we created a bash script to help us running migrations and studio locally.

```bash
#!/usr/bin/env sh

SECRET_ARN="..."

SECRET_BLOB=$(aws secretsmanager get-secret-value --output=text --query SecretString --secret-id "$SECRET_ARN")

PGUSER=$(node -pe 'JSON.parse(require("fs").readFileSync("/dev/stdin").toString()).username' <<< "$SECRET_BLOB")
PGPASSWORD=$(node -pe 'encodeURIComponent(JSON.parse(require("fs").readFileSync("/dev/stdin").toString()).password)' <<< "$SECRET_BLOB")
PGHOST=$(node -pe 'JSON.parse(require("fs").readFileSync("/dev/stdin").toString()).host' <<< "$SECRET_BLOB")
PGPORT=$(node -pe 'JSON.parse(require("fs").readFileSync("/dev/stdin").toString()).port' <<< "$SECRET_BLOB")
PGDATABASE=$(node -pe 'JSON.parse(require("fs").readFileSync("/dev/stdin").toString()).dbname' <<< "$SECRET_BLOB")

export DATABASE_URL="postgresql://$PGUSER:$PGPASSWORD@$PGHOST:$PGPORT/$PGDATABASE"

exec "$@"
```

**Comment by vkro** [comment]
@janpio any update on if a solution for this might be forthcoming? We're deploying to GKE (so also using Docker), and using GSM to manage secrets, also not wanting to keep our db credentials in env variables.

**Comment by janpio** [comment]
Any update would be posted in this issue.

**Comment by ajota-admios** [comment]
It would definitely be helpful to know if this is being considered to prepare for it or not hold our breath.

**Comment by hiagoteixeira** [comment]
Any update on this issue, i'm troubling the same problem?

**Comment by saumyadip1782** [comment]
any update on this ?

**Comment by ThaFishDance** [comment]
Any update on this?

Im tried calling the Prisma client in my handler using my AWS Secret Manager function, but i am getting a schema.prisma file not found when separating the schema.prisma and the client

**Comment by mouhannad-sh** [comment]
An async callback to fetch the database url would be awesome for Prisma running on serverless functions like AWS Lambda.
clients like Knex have this as well 
https://knexjs.org/guide/#configuration-options

**Comment by ikoichi** [comment]
+1 especially for the AWS Lambda scenario.

**Comment by arashi-dev** [comment]
+1 It would be great if we could have something like `FILE()` just like `ENV()` function. or maybe an approach to create a custom function as well and use it anywhere... (something like an extension or plugin to add custom functionality to the schema)

**Comment by salouri** [comment]
We had to drop using Prisma for all of our services because it lacks this functionality....

**Comment by NiallJoeMaher** [comment]
I'm about to drop Prisma because of this too.

**Comment by arashi-dev** [comment]
> I'm about to drop Prisma because of this too.

I think you can make a workaround, by creating a bash script to read the secret, set it as env, and execute the prisma cli command. just like what this person did in this comment (in the current issue):
https://github.com/prisma/prisma/issues/7534#issuecomment-1047745694

or if you can also get help from ChatGPT :)

**Comment by floelhoeffel** [comment]
We are currently looking into the design of a library that gives access to Prisma Migrate from code without depending on the CLI. If this sounds relevant to you, please help us with this 5 min survey so we can learn more about your specific needs.

https://prisma103696.typeform.com/to/btmMPJH6

🙇 Thank you very much!

**Comment by bnakhlis** [comment]
any progress on this?

**Comment by PhentomPT** [comment]
In my case we have a **_long time running process_** that should be able to get the DB url according to the latest secrets.
Meaning if a query failed because the secrets for the DB have refreshed we needed to retry a connection and repeat the operation.

Here's what worked for our case:
```typescript
import { PrismaClient, Prisma } from "@prisma/client";
import { getDbUrl } from "@/services/credential-manager";

export * from "@prisma/client";

let globalPrisma: PrismaClient;
const globalPrismaReconnect = {
  counter: 0,
  attempts: 2 as const,
};

async function newClient() {
  globalPrisma = new PrismaClient({
    datasources: {
      db: {
        url: await getDbUrl(),
      },
    },
  }).$extends({
    query: {
      async $allOperations({ model, operation, args, query }) {
        try {
          const result = await query(args);
          globalPrismaReconnect.counter = 0;
          return result;
        } catch (e) {
          const isPrismaInitError = e instanceof Prisma.PrismaClientInitializationError;

          if (isPrismaInitError && globalPrismaReconnect.counter > globalPrismaReconnect.attempts) {
            globalPrismaReconnect.counter = 0;
            console.error("DB reconnect failure");
            throw new Error(e.message);
          }

          if (!isPrismaInitError) {
            throw new Error(e);
          }

          // reconnect
          console.log("DB reconnecting...");
          globalPrismaReconnect.counter++;
          await newClient();

          // retry query with new connection
          return globalPrisma[model][operation](args);
        }
      },
    },
  }) as PrismaClient;
}

(async () => {
  await newClient();
})();

export const prisma = () => globalPrisma;
```

With the caveat that when using prisma we need to:
```typescript
import { prisma } from "@/services/db";

// instead of prisma.<collection>.<operation>
prisma().<collection>.<operation>
```

**Comment by visomi-dev** [comment]
This is my code for nestjs service:

```typescript
import { Injectable, OnModuleInit } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import {
  SecretsManagerClient,
  GetSecretValueCommand,
} from '@aws-sdk/client-secrets-manager';

import { PrismaClient } from '@prisma/client';

@Injectable()
export class PrismaService implements OnModuleInit {
  constructor(private readonly configService: ConfigService) {}

  secretManager = new SecretsManagerClient();

  client = new PrismaClient();

  async onModuleInit() {
    const { SecretString: secrets } = await this.secretManager.send(
      new GetSecretValueCommand({
        SecretId: this.configService.get('DATABASE_SECRET_NAME'),
      }),
    );

    const { host, port, username, password, database } = JSON.parse(secrets);

    this.client = new PrismaClient({
      datasources: {
        db: {
          url: `mysql://${username}:${password}@${host}:${port}/${database}`,
        },
      },
    });

    await this.client.$connect();
  }
}
```

**Comment by np-kyokyo** [comment]
We are currently facing the same issue. In our project, we manage the API with serverless lambda and connect to AWS RDS from there, using multiple databases. 
The secrets needed for this connection are managed by the Secrets Manager. Therefore, when executing commands like prisma migrate against AWS RDS, since dotenv can only set static values, we need to dynamically retrieve values from the Secrets Manager, set the environment variables appropriately, and then execute the prisma migrate command. 
Additionally, because we're using multiple databases, we need to specify the URL when creating the client. Having two ways to specify the URL is quite confusing for us. 

If possible, it would be ideal to centralize the URL specification in a configuration file and have an interface in the configuration file that can execute TypeScript. For example, I think it would be best to prepare a file like prisma.config.ts, similar to serverless.ts, and not describe the settings in the .prisma schema file. It would be great if we could pass dotenv variables to config.ts and branch the settings.

**Comment by OwaisAD** [comment]
Any updates on this after 2+ years....?

**Comment by mateussaggin** [comment]
We are currently struggling with the extra effort because we don't have this feature.

**Comment by ibz096** [comment]
Currently I create a new `PrismaClient` for an existing database with schema already defined like so:
```javascript
const prisma = new PrismaClient({
    datasourceUrl: await azure.fetchSecretFromAzureKeyVault('SQLCONNSTR-DATABASE-URL')
});
```
, but I need this to work during the migration/deployment to generate the DB schema

**Comment by felipecornejo1** [comment]
Also looking for this kind of solution to fetch db credentials from aws secrets manager @janpio

It'd be really nice to have this implemented

**Comment by SourenaSahraian** [comment]
It's surprising that this issue, despite being open for almost three years and receiving numerous requests for a feature that seems basic and widely needed, has yet to see a resolution from the Prisma team.

**Comment by ikoichi** [comment]
This is how I partially solved the issue:

```javascript
import { PrismaClient } from "@prisma/client";
import { getSecretValue } from "./secrets";

export const getPrismaClient = async () => {
  const secretName = "my-secret-key";
  const connectionStringSecret = await getSecretValue(secretName || ""); // get the secret from AWS

  const connectionString = connectionStringSecret[secretName];

  const client = new PrismaClient({
    datasources: {
      db: {
        url: connectionString,
      },
    },
    log: ["info", "warn", "error"],
  });

  return client;
};
```

Both for migrations / schema push I still use the connection_string in plain text on the `.env` file on my local machine

**Comment by johnrusch** [comment]
Yeah, this really brought my development to a halt unfortunately.

Manually overriding the datasources in the initiation of the PrismaService doesn't work for me sometimes. Also, when I'm developing locally I have to comment out the manual override. 

Commenting so I can get a notification about any updates on this.

**Comment by johnrusch** [comment]
> Yeah, this really brought my development to a halt unfortunately.
> 
> Manually overriding the datasources in the initiation of the PrismaService doesn't work for me sometimes. Also, when I'm developing locally I have to comment out the manual override.
> 
> Commenting so I can get a notification about any updates on this.

Okay, I got some rest and I got the manual datasource override working consistently.

I was trying to use NestJS's config module before to do some of their magic in hopes that it would give the Prisma client the database connection string more predictably and cleanly. I ended up just referencing env variables when creating the Prisma client (in my case in my prisma service) similar to @ikoichi.

```
@Injectable()
export class PrismaService
  extends PrismaClient
  implements OnModuleInit, OnModuleDestroy
{
  constructor() {
    const {
      DB_USERNAME,
      DB_PASSWORD,
      DB_HOST,
      DB_PORT,
    } = process.env;
    const dbUrl = `postgres://${DB_USERNAME}:${DB_PASSWORD}@${DB_HOST}:${DB_PORT}/mydatabase`;
    super({
      datasources: {
        db: {
          url: dbUrl,
        },
      },
    });
  }
  async onModuleInit() {
    await this.$connect();
  }

  async onModuleDestroy() {
    await this.$disconnect();
  }
}
```

**Comment by np-kyokyo** [comment]
I am trying to address this issue using the bun shell, like this:

```
import { $ } from "bun";
import { getDatabseUrl } from "../utils/get-databse-url";

const url = getDatabseUrl(); // get credentials from AWS Secret Manager client

await $`DATABASE_URL=${url} bun prisma db migrate deploy`;
```

I think it is much easier than using the AWS CLI.

**Comment by alist** [comment]
Linking https://github.com/prisma/prisma/pull/20740 since this in conjunction with @ikoichi's latest post seems to mostly solve the problem referenced in this issue. 

> This is how I partially solved the issue:
> 
> ```js
> import { PrismaClient } from "@prisma/client";
> import { getSecretValue } from "./secrets";
> 
> export const getPrismaClient = async () => {
>   const secretName = "my-secret-key";
>   const connectionStringSecret = await getSecretValue(secretName || ""); // get the secret from AWS
> 
>   const connectionString = connectionStringSecret[secretName];
> 
>   const client = new PrismaClient({
>     datasources: {
>       db: {
>         url: connectionString,
>       },
>     },
>     log: ["info", "warn", "error"],
>   });
> 
>   return client;
> };
> ```
> 
> Both for migrations / schema push I still use the connection_string in plain text on the `.env` file on my local machine

People can use a .env file for migrations. This is how I do it:

```bash
dotenv -c production -- npx prisma migrate deploy
```

**Comment by huangyanyang666** [comment]
> Linking #20740 since this in conjunction with @ikoichi's latest post seems to mostly solve the problem referenced in this issue.
> 
> > This is how I partially solved the issue:
> > ```js
> > import { PrismaClient } from "@prisma/client";
> > import { getSecretValue } from "./secrets";
> > 
> > export const getPrismaClient = async () => {
> >   const secretName = "my-secret-key";
> >   const connectionStringSecret = await getSecretValue(secretName || ""); // get the secret from AWS
> > 
> >   const connectionString = connectionStringSecret[secretName];
> > 
> >   const client = new PrismaClient({
> >     datasources: {
> >       db: {
> >         url: connectionString,
> >       },
> >     },
> >     log: ["info", "warn", "error"],
> >   });
> > 
> >   return client;
> > };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Both for migrations / schema push I still use the connection_string in plain text on the `.env` file on my local machine
> 
> People can use a .env file for migrations. This is how I do it:
> 
> ```shell
> dotenv -c production -- npx prisma migrate deploy
> ```
If I dont have the access to know the production db url, how can I write it in env file?

**Comment by fliffdan** [comment]
It's 2024 and this basic functionality that exists in every other ORM still doesn't exist.

I'm leaving this comment to let you know I'm no longer using the library. Thanks!

**Comment by tmokmss** [comment]
This feature is also highly desired for Aurora DSQL, which requires to dynamically issue an auth token as password string. 
https://docs.aws.amazon.com/aurora-dsql/latest/userguide/authentication-authorization.html

**Comment by cnorthfield** [comment]
We are looking to migrate from TypeORM to Prisma, and this appears to be a blocker. Our use case is the AWS SSM Parameter Store; we dynamically fetch secure parameters at runtime for secure configuration.

**Comment by rickgonzalez** [comment]
This really hurts. I love using Prisma and we get tons of benefits from using it - such as the ability to package up our schema and manage the versioning of the database as we tend to make many changes during development.  Unfortunately when we do this, the **schema.prisma file demands a fixed env file** - and there isn't a good way to get around it.  The Schema is used for us - not in just migrations - but it is what defines our database connection when we include the package with our application.  I don't have a problem using secrets manager with many other application - just Prisma.  It seems that the code that was used to bake in the env file needs to be broken into and implement what this fellow envisioned four years ago.  With all the data breaches, it would seem that you guys would prioritize compatibility with secrets managers - as the more I have to do to get around this with scripts and what not, the less secure the data secrets are. I do really appreciate the hard work and almost magic that happens with this product. It has made my life easier over several projects - I would be happy to work on this or test if you all pick this one up.


```
datasource db {
  provider = "postgresql"
  _url      = await getDatabaseUrl();_
}
```

**Comment by shadoath** [comment]
@SevInf @jkomyno @jharrell @aqrln or @nikolasburk, could one of you kindly let us know if there are plans to work on this? Or how do you all handle this in a production environment?

**Comment by jharrell** [comment]
@shadoath @Erry91 this can be resolved using the new [prisma.config.ts file](https://www.prisma.io/docs/orm/reference/prisma-config-reference) and [dynamically setting a datasourceUrl](https://www.prisma.io/docs/orm/reference/prisma-client-reference#datasourceurl) when you instantiate Prisma Client. 

The Prisma config would be for migrations and querying would be handled by the datasource URL. Is there any other missing functionality?

**Comment by azamuddin** [comment]
> [@shadoath](https://github.com/shadoath) [@Erry91](https://github.com/Erry91) this can be resolved using the new [prisma.config.ts file](https://www.prisma.io/docs/orm/reference/prisma-config-reference) and [dynamically setting a datasourceUrl](https://www.prisma.io/docs/orm/reference/prisma-client-reference#datasourceurl) when you instantiate Prisma Client.
> 
> The Prisma config would be for migrations and querying would be handled by the datasource URL. Is there any other missing functionality?

@jharrell actually, using Aurora DSQL for example, running migrations would need overriding datasource URL as well, so migrations & querying should use the dynamic datasource URL.

**Comment by igalklebanov** [comment]
Hey 👋 

The new `prisma.config.ts` lets you use whatever you want to retrieve your secrets with.
